openwisp_controller/config/whois/service.py

```diff
@@ -0,0 +1,283 @@
+        with DeviceChecksumView.
+        """
+        OrganizationConfigSettings = load_model("config", "OrganizationConfigSettings")
+        Config = load_model("config", "Config")
+
+        org_id = self.device.organization.pk
+        org_settings = cache.get(self.get_cache_key(org_id=org_id))
+        if org_settings is None:
+            try:
+                org_settings = OrganizationConfigSettings.objects.get(
+                    organization=org_id
+                )
+            except OrganizationConfigSettings.DoesNotExist:
+                # If organization settings do not exist, fall back to global setting
+                return app_settings.WHOIS_ENABLED
+            cache.set(
+                self.get_cache_key(org_id=org_id),
+                org_settings,
+                timeout=Config._CHECKSUM_CACHE_TIMEOUT,
+            )
+        return getattr(org_settings, "whois_enabled", app_settings.WHOIS_ENABLED)
+
+    def _need_whois_lookup(self, new_ip):
+        """
+        This is used to determine if the WHOIS lookup should be triggered
+        when the device is saved.
+
+        The lookup is not triggered if:
+            - The new IP address is None or it is a private IP address.
+            - The WHOIS information of new ip is already present.
+            - WHOIS is disabled in the organization settings. (query from db)
+        """
+
+        # Check cheap conditions first before hitting the database
+        if not self.is_valid_public_ip_address(new_ip):
+            return False
+
+        if self._get_whois_info_from_db(new_ip).exists():
+            return False
+
+        return self.is_whois_enabled
+
+    def get_device_whois_info(self):
+        """
+        Used to get WhoIsInfo for a device if last_ip is valid public ip
+        and WhoIs is enabled.
+        """
+        ip_address = self.device.last_ip
+        if not (self.is_valid_public_ip_address(ip_address) and self.is_whois_enabled):
+            return None
+
+        return self._get_whois_info_from_db(ip_address=ip_address).first()
+
+    def trigger_whois_lookup(self):
+        """
+        Trigger WhoIs lookup based on the conditions of `_need_whois_lookup`.
+        Task is triggered on commit to ensure redundant data is not created.
+        """
+        if self._need_whois_lookup(self.device.last_ip):
+            transaction.on_commit(
+                lambda: self.fetch_whois_details.delay(
+                    device_pk=self.device.pk,
+                    initial_ip_address=self.device._initial_last_ip,
+                    new_ip_address=self.device.last_ip,
+                )
+            )
+
+    # device_pk is used when task fails to report for which device failure occurred
+    @shared_task(
+        bind=True,
+        base=WhoIsCeleryRetryTask,
+        **app_settings.API_TASK_RETRY_OPTIONS,
+    )
+    def fetch_whois_details(self, device_pk, initial_ip_address, new_ip_address):
```

Comment:
Having tasks here is not really consistent with the rest of the openwisp code base. We use to group celery tasks in a file named `tasks.py`. Can you follow that pattern and maintain consistencyy please?